### PR TITLE
[PASS] Add GPU IR verifier

### PR DESCRIPTION
**Pull request by merrymercy**
Add a pass to check whether a cuda ir is valid

### CODE REVIEW

**Comment by merrymercy** [comment]
```
/workspace/src/arithmetic/modular.cc:168:1: fatal error: error writing to /tmp/cc30MweO.s: No space left on device
```

File: include/tvm/ir_pass.h [hunk 1]
```diff
@@ -477,6 +477,22 @@ LoweredFunc LowerIntrin(LoweredFunc f, const std::string& target);
...
+ * \brief Verify the correctness of a cuda ir
+ *        It will check the whether the amount of shared memory or the number of threads
+ *        in a block exceeds the limit
+ * \param stmt The statement to be checked
+ * \param max_shared_memory_per_block Total amount of shared memory per block (in bytes).
+ * \param max_thread_per_block Maximum number of threads per block.
+ * \return valid Whether it is a valid cuda ir
+ *
+ */
+bool VerifyCuda(Stmt stmt,
```
Comment (tqchen) [inline include/tvm/ir_pass.h]: This is CUDA specific? or should we call it VerifyGPUCode

**Comment by tqchen** [comment]
@merrymercy please act on the comments and fix the ci error

File: src/pass/verify_gpu_code.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#include <tvm/api_registry.h>
+#include <tvm/ir.h>
+#include <tvm/ir_visitor.h>
+
+namespace tvm {
+namespace ir {
+
+class GPUCodeVerifier : public IRVisitor {
+ public:
+  bool verify(tvm::Stmt stmt, int max_shared_memory_per_block, int max_thread_per_block) {
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: CamelCase for fucnctions

File: src/pass/verify_gpu_code.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+
+  void Visit_(const Allocate *op) {
+    IRVisitor::Visit_(op);
+    // visit an allocation of a buffer in shared memory, record its size
+    if (shared_buffers_.count(op->buffer_var.get()) != 0) {
+      int64_t size = op->type.bytes();
+      for (auto dim : op->extents) {
+        size *= dim.as<IntImm>()->value;
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: op-> constant_allocation_size()

File: src/pass/verify_gpu_code.cc [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      }
+    } else if (op->attr_key == attr::thread_extent) {
+      VarExpr var = op->node.as<tvm::IterVarNode>()->var;
+      const auto *extent = op->value.as<IntImm>();
+      CHECK(extent);
+
+      // record the number of threads in a block
+      std::string name = var.get()->name_hint;
+      if (name == "threadIdx.x" || name == "threadIdx.y" || name == "threadIdx.z") {
+        if (visited_threads_.find(name) == visited_threads_.end()) {
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: !count(name)

File: src/pass/verify_gpu_code.cc [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  std::unordered_set<const tvm::Variable *> shared_buffers_;
+  std::unordered_set<std::string> visited_threads_;
+  size_t shared_memory_per_block_;
+  size_t thread_per_block_;
+
+  size_t max_shared_memory_per_block_;
+  size_t max_thread_per_block_;
+
+  bool valid{true};
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: valid_

File: src/pass/verify_gpu_code.cc [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+  void reset_() {
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: Reset()

Comment (tqchen) [reply src/pass/verify_gpu_code.cc]: If this visitor is only used once, reset is not necessary

File: src/pass/verify_gpu_code.cc [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  void Visit_(const ProducerConsumer *op) {
+    if (nest_level_ == 0) {
+      // enter a new kernel, reset statistics
+      reset_();
+    }
+
+    if (op->is_producer) {
+      nest_level_++;
```
Comment (tqchen) [inline src/pass/verify_gpu_code.cc]: prefer --nest_+level

**Comment by tqchen** [comment]
@eqy can you also do a round of codereview?

**Comment by eqy** [comment]
@tqchen @merrymercy if we also save the number of threads per dimension (x, y, z), perhaps we can also use this to capture
`CL_INVALID_WORK_ITEM_SIZE` (e.g., in https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clEnqueueNDRangeKernel.html)
for OpenCL kernels. But I am not sure if this can happen in CUDA land, so we could also use a separate verifier pass there.

**Comment by tqchen** [comment]
It would be a good idea to pass in as many constraints as possible and allowing defaults to non-constraints. One possible way to do so is to allow pass in of a ```Map<str, value>``` so we don't have a lot of positional arguments

File: src/pass/verify_gpu_code.cc (hunk 6 above)
Comment (merrymercy) [reply src/pass/verify_gpu_code.cc]: reset is needed because there might be several gpu kernels in one statement.

**Comment by merrymercy** [comment]
ready for review

**Comment by eqy** [comment]
Is the plan to skip checking threadId/workitem dimensions in this round?

**Comment by tqchen** [comment]
please fix the compiler warning http://mode-gpu.cs.washington.edu:8080/blue/organizations/jenkins/dmlc%2Ftvm/detail/PR-1296/11/pipeline

Currently we set compiler warning as error so built won't pass if there is a warning

**Comment by merrymercy** [comment]
@eqy your comment are addressed
@tqchen  no space left on CI

File: tests/python/unittest/test_pass_verify_gpu_code.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
+"""Test gpu code verifier"""
+import tvm
+
+global valid
```
Comment (tqchen) [inline tests/python/unittest/test_pass_verify_gpu_code.py]: always avoid use of global variable, to carry state, you can use a closure to capture a list

**Comment by tqchen** [comment]
The test error likely indicate there is some problem with the current PR when importing runtime only dll.

**Pull request merged**